code/modules/holomap/holomap.dm

```diff
@@ -0,0 +1,500 @@
+		region size it will be centered in the `bigassicon`. Happy Viewing.
+	*/
+	var/region_width = (region_selection[3] - region_selection[1]) + 1
+	var/region_height = (region_selection[4] - region_selection[2]) + 1
+
+	var/region_x_start = (region_width > 240) ? 9 : region_selection[1]
+	var/region_y_start = (region_height > 240) ? 9 : region_selection[2]
+
+	var/region_x_finish = (region_width > 240) ? region_x_start + 239 : region_selection[3]
+	var/region_y_finish = (region_width > 240) ? region_y_start + 239 : region_selection[4]
+
+	var/render_x_start = (region_width > 240) ? -8 : 128 - (region_width / 2)
+	var/render_y_start = (region_width > 240) ? -8 : 128 - (region_height / 2)
+
+	var/icon_to_use = 'icons/effects/holomap_parts.dmi'
+
+	for(var/x_index in region_x_start to region_x_finish)
+		for(var/y_index in region_y_start to region_y_finish)
+			var/turf/T = locate(x_index, y_index, zlevel_rendered)
+			var/area/A = get_area(T)
+			if(A.type != /area/space && A.type != /area/space/nearstation)
+				if(location_areas[A])
+					location_areas[A][2] += T
+				else
+					location_areas[A] = list(A, list(T))
+			var/W = FALSE
+			for(var/obj/structure/window/full/WW in T.contents)
+				W = TRUE
+				break
+			for(var/obj/effect/spawner/window/WW in T.contents)
+				W = TRUE
+				break
+
+			/*
+				This works by using 2x2 sprites from 'icons/effects/holomap_parts.dmi' to construct the holomap;
+				We use different colors that we will swap later on for floors and walls (BLACK-->floor;WHITE-->wall)
+				The walls can either be a full 2x2 pixel box, 1x2 pixel wall, or a 3/4 pixel corner depending on their surroundings...
+				All the following logic is to determine what icon to use.
+			*/
+			//this is what all the logic is for, setting this variable
+			var/icon_state_to_use
+			//as well as this since it modifies the wall sprite
+			var/dir_to_use = 2
+
+			//the following variables are set conditionally on the state of the turf
+			//so we can render the right sprite for this turf
+			var/neighbor_space = FALSE
+			var/list/neighbors_alien = list()
+			var/list/neighbors_solid = list()
+
+	//--------------{{{ DISCOVERY PHASE }}}------------------------------
+
+			for(var/i in GLOB.cardinal)
+				var/turf/found_turf = get_step(T, i)
+				var/area/found_area = get_area(found_turf)
+				var/found_window = FALSE
+
+				for(var/obj/structure/window/full/WW in found_turf.contents)
+					found_window = TRUE
+					break
+				for(var/obj/effect/spawner/window/WW in found_turf.contents)
+					found_window = TRUE
+					break
+
+				for(var/j in GLOB.diagonals)
+					var/turf/found_s_turf = get_step(T, j)
+					var/area/found_s_area = get_area(found_s_turf)
+					if(found_s_area == /area/space || found_s_area == /area/space/nearstation || isspaceturf(found_s_turf))
+						neighbor_space = TRUE
+						break
+
+				if(neighbor_space || found_area == /area/space || found_area == /area/space/nearstation || isspaceturf(found_turf))
```

Comment:
is there ever a case where a turf passes `isspaceturf()` and doesn't have `/area/space` or `/area/space/nearstation` as its area

Reply:
sometimes we extend a station area into space to power injectors or disposal machines, and on Shepard I have station areas on space turfs to power the space bridge buttons (might be the case on other stations with custom airlocks)